src/Features/LanguageServer/Protocol/Handler/SpellCheck/AbstractSpellCheckingHandler.cs

```diff
@@ -147,22 +148,62 @@ protected AbstractSpellCheckHandler()
             return result;
         }
 
-        private async Task<TReport> ComputeAndReportCurrentSpansAsync(
+        private async IAsyncEnumerable<TReport> ComputeAndReportCurrentSpansAsync(
             Document document,
             ISpellCheckSpanService service,
             string resultId,
-            CancellationToken cancellationToken)
+            [EnumeratorCancellation] CancellationToken cancellationToken)
         {
+            var textDocumentIdentifier = ProtocolConversions.DocumentToTextDocumentIdentifier(document);
 
             var text = await document.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
             var spans = await service.GetSpansAsync(document, cancellationToken).ConfigureAwait(false);
 
-            using var _ = ArrayBuilder<LSP.VSInternalSpellCheckableRange>.GetInstance(spans.Length, out var result);
+            // protocol requires the results be in sorted order
+            spans = spans.Sort(static (s1, s2) => s1.TextSpan.CompareTo(s1.TextSpan));
+
+            if (spans.Length == 0)
+            {
+                yield return CreateReport(textDocumentIdentifier, Array.Empty<int>(), resultId);
+                yield break;
+            }
+
+            // break things up into batches of 1000 items.  That way we can send smaller messages to the client instead
+            // of one enormous one.
+            const int chunkSize = 1000;
+            for (var batchStart = 0; batchStart < spans.Length; batchStart += chunkSize)
+            {
+                var batchEnd = Math.Min(batchStart + chunkSize, spans.Length);
+                var batchSize = batchEnd - batchStart;
+
+                // Each span is encoded as a triple of ints.  The 'kind', the 'relative start', and the 'length'.
+                // 'relative start' is the absolute-start for the first span, and then the offset from the end of the
+                // last span for all others.
+                var triples = new int[batchSize * 3];
+                var triplesIndex = 0;
+                var lastSpanEnd = 0;
```

Comment:
As talked about it privately, there's a mistake in the interpretation of the protocol. The start position should be the relative offset from the end of the previous span, regardless of whether that span is the same range or a prior range.
I think the fix is simply to move this line above line 174.